### PR TITLE
Fix to cuda tensor

### PR DESCRIPTION
**Pull request by tscholak**
I had to comment the `TensorLike` instance for `Tensor`, because it would allow for switching gradients off or on from outside of `IO` which mustn't be allowed.

### CODE REVIEW

File: experimental/gradually-typed/src/Torch/GraduallyTyped/Tensor/Type.hs [hunk 1]
```diff
@@ -242,26 +242,16 @@ withGradient ::
...
-
-instance SSetGradient ('Gradient 'WithoutGradient) where
-  sSetGradient (SGradient SWithoutGradient) tensor =
-    cast2 ATen.tensor_set_requires_grad_b tensor False
+-- | Turn gradient computations off or on for a tensor.
+sSetGradient ::
+  forall gradient gradient' layout device dataType shape.
+  SGradient gradient ->
+  Tensor gradient' layout device dataType shape ->
+  IO (Tensor gradient layout device dataType shape)
```
Comment (jul1u5) [inline experimental/gradually-typed/src/Torch/GraduallyTyped/Tensor/Type.hs:250]: Shouldn't the return type be wrapped in `MonadThrow` instead of `IO` here?

**Review by jul1u5 — COMMENTED** [review]
Oh, I missed this. Maybe we should wrap the return type of `fromTensor` with `MonadThrow`? I could make a PR for this.

**Comment by tscholak** [comment]
It has to be IO, unfortunately. (Or a special yet-to-be-defined TorchMonad that evaluates to IO.) The problem is that turning on or off gradients for an existing tensor changes the global state of the torch runtime. As a side effect, that runtime builds the compute graph in order to be able to backpropagate and compute gradients when requested. That messes with referential transparency.

**Review by jul1u5 — APPROVED** [review]
Oh, ok, that makes sense. I think we can merge this.

File: experimental/gradually-typed/src/Torch/GraduallyTyped/Tensor/Type.hs [hunk 2]
```diff
@@ -1404,17 +1386,19 @@ sToTensorRaw gradient' layout device x = do
       t <- UnsafeTensor <$> cast2 ATen.empty_lo dims' opts
       withTensor t $ \ptr ->
         tensorPokeElemOff ptr 0 dims' x
-      pure t
+      sSetDevice device t
```
Comment (tscholak) [inline experimental/gradually-typed/src/Torch/GraduallyTyped/Tensor/Type.hs:1389]: @jul1u5 fyi

File: experimental/gradually-typed/src/Torch/GraduallyTyped/Tensor/Type.hs [hunk 3]
```diff
@@ -1404,17 +1386,19 @@ sToTensorRaw gradient' layout device x = do
...
 fromTensorRaw ::
   forall gradient layout device a dType dims.
   (TensorLike a dType dims, TensorLikeRaw a, SGetDims dims) =>
   Tensor gradient layout device ('DataType dType) ('Shape dims) ->
   a
-fromTensorRaw t = unsafePerformIO $
-  withTensor t $ \ptr -> tensorPeekElemOff ptr 0 (fromInteger . dimSize <$> getDims t)
+fromTensorRaw t =
+  unsafePerformIO $
+    sSetDevice (SDevice SCPU) t
```
Comment (tscholak) [inline experimental/gradually-typed/src/Torch/GraduallyTyped/Tensor/Type.hs:1400]: @jul1u5 fyi
